reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,85 +83,116 @@ object StreamSubscriber {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
+          case WaitingOnUpstream(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
           case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
           case o         => o -> F.unit
         }
-        case OnDequeue(r) => {
-          case Uninitialized          => RequestBeforeSubscription(r) -> F.unit
-          case Idle(sub)              => WaitingOnUpstream(sub, r) -> F.delay(sub.request(1))
-          case err @ UpstreamError(e) => err -> r.complete(e.asLeft)
-          case UpstreamCompletion     => UpstreamCompletion -> r.complete(None.asRight)
-          case o                      => o -> r.complete((new Error(s"received request in invalid state [$o]")).asLeft)
+        case OnDequeue => {
+          case Uninitialized =>
+            RequestBeforeSubscription -> F.unit
+          case Idle(sub) =>
+            WaitingOnUpstream(sub) -> F.delay(sub.request(batchSize))
+          case NextBatch(sub) =>
+            Receiving(sub) -> F.delay(sub.request(batchSize))
+          case Receiving(sub) =>
+            Receiving(sub) -> F.unit
+          case err @ UpstreamError(e) =>
+            err -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case UpstreamCompletion =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case o =>
+            o -> (q.enqueue1((new Error(s"received request in invalid state [$o]")).asLeft) >> q
+              .enqueue1(None.asRight))
         }
       }
 
-    Ref.of[F, State](Uninitialized).map { ref =>
-      new FSM[F, A] {
-        def nextState(in: Input): F[Unit] = ref.modify(step(in)).flatten
-        def onSubscribe(s: Subscription): F[Unit] = nextState(OnSubscribe(s))
-        def onNext(a: A): F[Unit] = nextState(OnNext(a))
-        def onError(t: Throwable): F[Unit] = nextState(OnError(t))
-        def onComplete: F[Unit] = nextState(OnComplete)
-        def onFinalize: F[Unit] = nextState(OnFinalize)
-        def dequeue1: F[Either[Throwable, Option[A]]] =
-          Deferred[F, Out].flatMap(p => ref.modify(step(OnDequeue(p))).flatten >> p.get)
-      }
+    (Ref.of[F, State](Uninitialized), Ref.of[F, Long](0)).mapN {
+      case (ref, counter) =>
+        new FSM[F, A] {
+          def nextState(in: Input): F[Unit] =
+            counter.get.flatMap(c => ref.modify(step(in, c)).flatten)
+          def onSubscribe(s: Subscription): F[Unit] =
+            nextState(OnSubscribe(s))
+          def onNext(a: A): F[Unit] =
+            counter.update(_ + 1) >> nextState(OnNext(a))
+          def onError(t: Throwable): F[Unit] =
+            nextState(OnError(t))
+          def onComplete: F[Unit] =
+            nextState(OnComplete)
+          def onFinalize: F[Unit] =
+            nextState(OnFinalize)
+          def dequeue1: F[Chunk[Either[Throwable, Option[A]]]] =
```

Comment:
Instead of `dequeueChunk1` we could call `.stream` on the queue directly. But that makes testing with the reactive streams tck more difficult since we need a way to tell this impl to pull from the `Publisher` on command.